### PR TITLE
8256266: Binding variables don't correct support declaration annotations and the final modifier

### PR DESCRIPTION
**Pull request by lahodaj**
Basically, javac does not handle code like this properly:

```
import java.lang.annotation.*;

public class B {
     public void test(Object o) {
         if (o instanceof @A String s) {}
     }
}

@Target(ElementType.LOCAL_VARIABLE)
@interface A {}
```

The parser needs to allow declaration annotations on binding variables, see:
http://cr.openjdk.java.net/~gbierman/jep394/jep394-20201012/specs/patterns-instanceof-jls.html#jls-14.30.1.1

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256266](https://bugs.openjdk.java.net/browse/JDK-8256266): Binding variables don't correct support declaration annotations and the final modifier

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1192/head:pull/1192`
`$ git checkout pull/1192`

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java [hunk 1]
```diff
@@ -933,16 +933,29 @@ JCExpression term2Rest(JCExpression t, int minprec) {
...
                 JCTree pattern;
                 if (token.kind == IDENTIFIER) {
                     checkSourceLevel(token.pos, Feature.PATTERN_MATCHING_IN_INSTANCEOF);
-                    JCModifiers mods = F.at(Position.NOPOS).Modifiers(0);
                     JCVariableDecl var = toP(F.at(token.pos).VarDef(mods, ident(), type, null));
                     TreeInfo.getStartPos(var);
                     pattern = toP(F.at(typePos).BindingPattern(var));
                     TreeInfo.getStartPos(pattern);
                 } else {
+                    if (mods.annotations.nonEmpty()) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:946]: We don't seem to check for the source level here? E.g. is it possible that `instanceof @Foo Bar` will not issue the usual error if compiled with source != 16?

**Review by mcimadamore — COMMENTED** [review]
Changes look good - I just have a question on the parser code

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java (hunk 1 above)
Comment (lahodaj) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:950]: Oops. Having `instanceof @Foo Bar` is valid, due to type annotations. See e.g.:
https://github.com/openjdk/jdk/blob/d2b66196b4509a32613f476d7247f207289a75e3/test/langtools/tools/javac/annotations/typeAnnotations/TargetTypes.java#L102

But, when I was adding support for the final modifier on binding variables, I've forgot to disable the final modifier on non-pattern matching `instanceof`, so code like `o instanceof final String` was accepted, which is, of course, wrong.

This should be fixed by:
https://github.com/openjdk/jdk/pull/1192/commits/a737f9229adf86af90fdab98a67ad009d76ab44c

**Review by mcimadamore — APPROVED** [review]
Changes look good to me

**Comment by lahodaj** [comment]
Has been integrated as:
https://git.openjdk.java.net/jdk16/pull/78

**Pull request closed without merging**